### ISSUE TITLE
Query string flags

### ISSUE THREAD

**Issue by GlennS** [tiangolo/fastapi#610]
**Is your feature request related to a problem? Please describe.**

It is valid in HTML to have a query string parameter which has no value. For example: `?my-flag`.

I usually think of this as a shorthand for writing `?my-flag=true`.

FastAPI ignores these.

**Describe the solution you'd like**

FastAPI should not ignore these flags.

It could set their value to be:

 - `True`.
 - Empty string `''` (I think this is bad idea - in intent these are usually for turning features on/off).
 - Some constant like `FastAPI.params.EmptyFlag`.

Alternatively, you could make some special type which marks a query string parameter as present or not.

**Comment by Trim21** [comment]
It is valid in HTML to have a query string parameter which has no value. 

Did you mean HTTP request instead of HTML?

**Comment by Trim21** [comment]
I try to implement this but I found that it's a `starlette` issue.

```python
from starlette.requests import Request
from fastapi.params import Depends
from starlette.testclient import TestClient


class BoolFlagQuery(Depends):
    def __init__(self, name):
        super().__init__(dependency=self, use_cache=True)
        self.name = name

    async def __call__(self, request: Request) -> bool:
        print(request.query_params, type(request.query_params))
        # q=1 <class 'starlette.datastructures.QueryParams'
        value: str = request.query_params.get(self.name)
        return value is not None


from fastapi import FastAPI
app = FastAPI()


@app.get('/path')
async def handler(flag: bool = BoolFlagQuery('flag')):
    return flag


if __name__ == '__main__':
    with TestClient(app=app) as client:
        with client.get('http://example.com/path?flag&q=1') as res:
            print(res.text) # false
```
`starlette` is using `urllib.parse.parse_qsl` to parse query string, and default paramaters of `parse_sql` will omit query with empty value

**Comment by Trim21** [comment]
```python
import urllib.parse

print(urllib.parse.parse_qsl('flag&q=1', keep_blank_values=True))
# [('flag', ''), ('q', '1')]
print(urllib.parse.parse_qsl('flag&q=1', keep_blank_values=False))
# [('q', '1')]
```

**Comment by euri10** [comment]
> **Is your feature request related to a problem? Please describe.**
> 
> It is valid in HTML to have a query string parameter which has no value. For example: `?my-flag`.
> 
> I usually think of this as a shorthand for writing `?my-flag=true`.
> 
> FastAPI ignores these.
> 

I'm not sure I understand, the below shows FastAPI deals fine with an endpoint that has the `my_flag` (underscore) without value set, when no value is provided the default is passed ?

So when you say FastAPI ignores these do you mean FastAPI returns None and should instead return True (which is possible below changing the default value of my_flag to True) ?

```
import pytest
from async_asgi_testclient import TestClient
from fastapi import FastAPI, Query

app = FastAPI()


@app.get("/items")
async def read_items(q: str = Query(None), my_flag: bool = Query(None)):
    return {"q": q, "my_flag": my_flag}


@pytest.mark.asyncio
async def test_empty():
    client = TestClient(app)
    response = await client.get("/items?my_flag&q=1")
    assert response.status_code == 200
    assert response.json() == {"q": "1", "my_flag": None}

    response = await client.get("/items?my_flag=false&q=1")
    assert response.status_code == 200
    assert response.json() == {"q": "1", "my_flag": False}
```

**Comment by Trim21** [comment]
@euri10 He means there is no way to distinguish between 'no value' and not passing this query.

You can't know the client is issuing requests with `client.get("/items?q=1")` or `client.get("/items?my_flag&q=1")` except you parse query string in handler.

**Comment by GlennS** [comment]
Yes, you are correct that I mean 'HTTP' not 'HTML'.

And yes, my issue is that I can't distinguish between "/some-url?flag" and "/some-url" at all.

The normal use of this is for things that you want to turn on optionally, which the current behaviour doesn't allow. Of course, it is not a big hassle to say that clients of my API must use `/some-url?flag=true` instead.

**Comment by GlennS** [comment]
I've opened https://github.com/encode/starlette/issues/670

**Comment by GlennS** [comment]
Starlette have made a change to address this: https://github.com/encode/starlette/pull/672/

Does this help?

**Comment by dmontagu** [comment]
@GlennS Yes, once there is a new release of starlette, we can update fastapi to support it.

If you need access to this now, everything would probably work if you installed starlette directly from the git repo and just ignored the version conflicts (I don't *think* any breaking changes have happened since the last release, but I'm not sure).

**Comment by tiangolo** [comment]
Here's the thing, what is now supported in Starlette is to have empty query param flags. So, they are in the `request.query_params` dict, but with a value of `None`, because actually nothing was provided.

So, you could extract them if you use the request directly and check if the query parameter exists and if it is None: https://fastapi.tiangolo.com/advanced/using-request-directly/

But it wouldn't make sense to make them used by default and take them as `True` implicitly, as that would be inconsistent with all the other cases.

For example, all these would make `q` `have a value `False`:

`http://example.com?q=false`
`http://example.com?q=0`
`http://example.com?q=""`
`http://example.com?q=`

But if then suddenly this made `q` to have a value `True`:

`http://example.com?q`

That would probably be quite unexpected.

So, yes, you can implement it by using the request directly, but it wouldn't make sense as a default in FastAPI.

**Comment by GlennS** [comment]
Thanks for the good advice. I will bear that in mind next time I make something with FastAPI.

**Comment by moraesjeremias** [comment]
Although you do can create uris with empty query strings, no value query strings simply don't follow [RFC 8820](https://tools.ietf.org/html/rfc8820) and [RFC 3986](https://tools.ietf.org/html/rfc3986#section-3.4) as they may cause erroneous client assumptions.

FastAPI shouldn't infer which type it should manage for `my-flag` in resource `/resource?my-flag`
